File: internal/index/persistence.go
package index

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"

	"github.com/takara-ai/miru-code/internal/utils"
)

type PersistencePaths struct {
	Root          string
	BM25Index     string
	SemanticIndex string
	Chunks        string
	Metadata      string
}

func PersistencePathsFor(root string) PersistencePaths {
	return PersistencePaths{
		Root:          root,
		BM25Index:     filepath.Join(root, "bm25_index.json"),
		SemanticIndex: filepath.Join(root, "semantic_index"),
		Chunks:        filepath.Join(root, "chunks.json"),
		Metadata:      filepath.Join(root, "metadata.json"),
	}
}

func PathsExist(paths PersistencePaths) []string {
	checks := []struct {
		name string
		path string
	}{
		{"bm25", paths.BM25Index},
		{"semantic", filepath.Join(paths.SemanticIndex, "meta.json")},
		{"chunks", paths.Chunks},
		{"metadata", paths.Metadata},
	}
	missing := []string{}
	for _, check := range checks {
		if _, err := os.Stat(check.path); err != nil {
			missing = append(missing, check.name)
		}
	}
	return missing
}

func SaveBM25(index *BM25Index, path string) error {
	data, err := json.Marshal(index.ToJSON())
	if err != nil {
		return err
	}
	return os.WriteFile(path, data, 0o644)
}

func LoadBM25(path string) (*BM25Index, error) {
	data, err := os.ReadFile(path)
	if err != nil {
		return nil, err
	}
	var payload BM25JSON
	if err := json.Unmarshal(data, &payload); err != nil {
		return nil, err
	}
	return BM25FromJSON(payload), nil
}

func SaveSemantic(index SemanticIndex, path string) error {
	switch idx := index.(type) {
	case *QuantizedVectorIndex:
		return idx.Save(path)
	case *VectorIndex:
		return idx.Save(path)
	default:
		return fmt.Errorf("Unsupported semantic index type")
	}
}

func LoadSemantic(path string) (SemanticIndex, error) {
	meta, err := readMeta(path)
	if err != nil {
		return nil, err
	}
	if meta.Storage == "int8" {
		return LoadQuantizedVectorIndex(path)
	}
	return LoadVectorIndex(path)
}

type IndexBundle struct {
	BM25     *BM25Index
	Semantic SemanticIndex
	Chunks   []utils.Chunk
	Metadata map[string]any
}

func SaveIndexBundle(paths PersistencePaths, bundle IndexBundle) error {
	if err := os.MkdirAll(paths.Root, 0o755); err != nil {
		return err
	}
	if err := SaveBM25(bundle.BM25, paths.BM25Index); err != nil {
		return err
	}
	if err := SaveSemantic(bundle.Semantic, paths.SemanticIndex); err != nil {
		return err
	}
	metadata := map[string]any{}
	for key, value := range bundle.Metadata {
		metadata[key] = value
	}
	metadata["vector_storage"] = string(SemanticStorageOf(bundle.Semantic))
	metadataData, err := json.Marshal(metadata)
	if err != nil {
		return err
	}
	if err := os.WriteFile(paths.Metadata, metadataData, 0o644); err != nil {
		return err
	}

	chunkData := make([]map[string]any, len(bundle.Chunks))
	for i, chunk := range bundle.Chunks {
		chunkData[i] = utils.ChunkToDict(chunk)
	}
	chunksJSON, err := json.Marshal(chunkData)
	if err != nil {
		return err
	}
	return os.WriteFile(paths.Chunks, chunksJSON, 0o644)
}

func LoadIndexBundle(root string) (IndexBundle, error) {
	paths := PersistencePathsFor(root)
	bm25, err := LoadBM25(paths.BM25Index)
	if err != nil {
		return IndexBundle{}, err
	}
	semantic, err := LoadSemantic(paths.SemanticIndex)
	if err != nil {
		return IndexBundle{}, err
	}
	chunkRaw, err := os.ReadFile(paths.Chunks)
	if err != nil {
		return IndexBundle{}, err
	}
	var chunkDicts []map[string]any
	if err := json.Unmarshal(chunkRaw, &chunkDicts); err != nil {
		return IndexBundle{}, err
	}
	chunks := make([]utils.Chunk, len(chunkDicts))
	for i, data := range chunkDicts {
		chunks[i] = utils.ChunkFromDict(data)
	}
	metadataRaw, err := os.ReadFile(paths.Metadata)
	if err != nil {
		return IndexBundle{}, err
	}
	var metadata map[string]any
	if err := json.Unmarshal(metadataRaw, &metadata); err != nil {
		return IndexBundle{}, err
	}
	return IndexBundle{BM25: bm25, Semantic: semantic, Chunks: chunks, Metadata: metadata}, nil
}

func SemanticIndexMatchesStorage(path string) bool {
	meta, err := readMeta(path)
	if err != nil {
		return false
	}
	stored := SemanticFloat32
	if meta.Storage == "int8" {
		stored = SemanticInt8
	}
	return stored == ResolveSemanticStorage()
}

File: src/index/persistence.ts
import { join } from "node:path";
import { BM25Index } from "./bm25.ts";
import { VectorIndex } from "./dense.ts";
import { QuantizedVectorIndex } from "./quantize.ts";
import type { SemanticIndex } from "./semantic-index.ts";
import { resolveSemanticStorage, semanticStorageOf } from "./vector-storage.ts";

export interface PersistencePaths {
  root: string;
  bm25Index: string;
  semanticIndex: string;
  chunks: string;
  metadata: string;
}

export function persistencePaths(root: string): PersistencePaths {
  return {
    root,
    bm25Index: join(root, "bm25_index.json"),
    semanticIndex: join(root, "semantic_index"),
    chunks: join(root, "chunks.json"),
    metadata: join(root, "metadata.json"),
  };
}

export async function pathsExist(paths: PersistencePaths): Promise<string[]> {
  const missing: string[] = [];
  const checks: [string, string][] = [
    ["bm25", paths.bm25Index],
    ["semantic", `${paths.semanticIndex}/meta.json`],
    ["chunks", paths.chunks],
    ["metadata", paths.metadata],
  ];
  for (const [name, p] of checks) {
    if (!(await Bun.file(p).exists())) {
      missing.push(name);
    }
  }
  return missing;
}

export async function saveBm25(index: BM25Index, path: string): Promise<void> {
  await Bun.write(path, JSON.stringify(index.toJSON()));
}

export async function loadBm25(path: string): Promise<BM25Index> {
  return BM25Index.fromJSON(JSON.parse(await Bun.file(path).text()));
}

export async function saveSemantic(index: SemanticIndex, path: string): Promise<void> {
  if (index instanceof QuantizedVectorIndex) {
    await index.save(path);
    return;
  }
  if (index instanceof VectorIndex) {
    await index.save(path);
    return;
  }
  throw new Error("Unsupported semantic index type");
}

export async function loadSemantic(path: string): Promise<SemanticIndex> {
  const meta = JSON.parse(await Bun.file(`${path}/meta.json`).text()) as {
    storage?: string;
  };
  if (meta.storage === "int8") {
    return QuantizedVectorIndex.load(path);
  }
  return VectorIndex.load(path);
}

export async function saveIndexBundle(options: {
  paths: PersistencePaths;
  bm25: BM25Index;
  semantic: SemanticIndex;
  chunks: unknown[];
  metadata: Record<string, unknown>;
}): Promise<void> {
  const { paths, bm25, semantic, chunks, metadata } = options;
  await saveBm25(bm25, paths.bm25Index);
  await saveSemantic(semantic, paths.semanticIndex);
  await Bun.write(
    paths.metadata,
    JSON.stringify({
      ...metadata,
      vector_storage: semanticStorageOf(semantic),
    }),
  );
  await Bun.write(paths.chunks, JSON.stringify(chunks));
}

export async function semanticIndexMatchesStorage(path: string): Promise<boolean> {
  const meta = JSON.parse(await Bun.file(`${path}/meta.json`).text()) as {
    storage?: string;
  };
  const stored = meta.storage === "int8" ? "int8" : "float32";
  return stored === resolveSemanticStorage();
}
